desktop-widgets/importgps.cpp

```diff
@@ -0,0 +1,329 @@
+	fileName(fileName), LocationUI(LocationUI)
+{
+	ui.setupUi(this);
+	connect(ui.timeDiffEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeDiffEditChanged()));
+	connect(ui.timeZoneEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeZoneEditChanged()));
+	connect(ui.timezone_backwards, SIGNAL(toggled(bool)), this, SLOT(changeZoneBackwards()));
+	connect(ui.timezone_forward, SIGNAL(toggled(bool)), this, SLOT(changeZoneForward()));
+	connect(ui.diff_backwards, SIGNAL(toggled(bool)), this, SLOT(changeDiffBackwards()));
+	connect(ui.diff_forward, SIGNAL(toggled(bool)), this, SLOT(changeDiffForward()));
+	connect(ui.GPSbuttonBox, SIGNAL(clicked(QAbstractButton *)), this, SLOT(buttonClicked(QAbstractButton *)));
+	coords.settingsDiff_offset = 0;
+	coords.timeZone_offset = 0;
+	coords.lat = 0;
+	coords.lon = 0;
+	pixmapSize = (int) (ui.diveDateLabel->height() / 2);
+}
+
+void  ImportGPS::buttonClicked(QAbstractButton *button)
+{
+	if (ui.GPSbuttonBox->buttonRole(button) == QDialogButtonBox::AcceptRole) {
+		// Write the coordinates in decimal degree format to the Coordinates QLineEdit of the LocationaInformationWidget UI:
+		LocationUI->diveSiteCoordinates->setText(QString::number(coords.lat) + ", " + QString::number(coords.lon));
+		LocationUI->diveSiteCoordinates->editingFinished();
+	} else {
+		close();
+	}
+}
+
+// Extract text from the present position in the file until
+//  the first 'delim' character is encountered.
+int ImportGPS::get_substring(FILE *f, char *buf, char delim)
+{
+	int j;
+	char c;
+	for (j = 0; (c = fgetc(f)); j++) {
+		if (c == delim) break;
+		if (c == EOF) return 1;
+		buf[j] = c;
+		if (j >= BUFFER_SIZE)
+			return 2;
+	}
+	buf[j] = 0x0; // terminate the string
+	return 0;
+}
+
+// Find the next occurence of a specified GPX element in the file,
+// characerised by a "<xxx " or "<xxx>" character sequence.
+// s specifies the name of the element searched for.
+// termc is the ending character of the element name search = '>' or ' '.
+int ImportGPS::find_xml_element(FILE *f, const char *s, char *buf, char termc)
+{
+	bool match = false;
+	int j;
+	char c;
+	char skipdelim = (termc == ' ') ? '>' : ' ';
+	do {
+		for (j = 0; (c = fgetc(f)); j++) {  // Find the first '<' character
+			if (c == '<')
+				break;
+			if (c == EOF)
+				return 1;
+			if (j >= BUFFER_SIZE)
+				return 2;
+		}
+		for (j = 0; (c = fgetc(f)); j++) {  // Extract the xml element name after the '<'
+			if (c == EOF)
+				return 1;
+			buf[j] = c;
+			if ((c == '>') || (c == ' '))
+				break;
+		}
+		buf[j] = 0x0;	// terminate the xml element name
+		if (!(strcmp(buf, "/trk"))) // End of GPS track found: return EOF
+			return 1;
+		if (c == skipdelim) continue; // if wrong delimiter for this element was found
+		if (strcmp(s, buf))       // Compare xml element name from gpx file with the
+			match = false;    // the element searched for.
+		else
+			match = true;
```

Comment:
can be `match = strcmp(s, buf) ? true : false`

Reply:
@bstoeger's comment taken for this one. The else part has been deleted now.